esp32: Integrate with the rp2 port mechanism to support user modules in cmake based builds.

mocleiri opened the pull request:
This work started as new work to try and restore the User Module support to the ESP32 port.  However now this work builds upon the rp2 port work which is working to upstreamed from their repository back into micropython master in pull request #6960.

Some of the work done here in py/usermod.cmake has been folded into #6960 so the changes in the most recent rebase are much simpler.

I've updated this initial message to better reflect the latest status of this pull request.

Using the latest changes you can build the firmware with user modules as follows:

```
$ cd ports/esp32

$ make USER_C_MODULES=/path/to/module/micropython.cmake 
```

stinos commented:
Can you update the relevant documentation as well so it is easier to to see what is needed? You only need a `make` step, not `cmake` ? Nitpick: code has superfluous whitespace additions and indentation is not consistent.

mocleiri reviewed (commented):
@stinos 
> Can you update the relevant documentation as well so it is easier to to see what is needed? 

Would the best place be the ESP32 port readme?

> You only need a `make` step, not `cmake` ? 

I added an option to pass through the additional parameters to the idf.py invocation.

I don't think cmake is called directly. Gnu make calls idf.py which calls cmake.

> Nitpick: code has superfluous whitespace additions and indentation is not consistent.

Thanks I'll remove the white space and fix the indentation on my next iteration.

This pull requests is more of a first step.  
To demonstrate a starting point that offers with custom firmwares could try out.

If any ones does and wants to tack on commits to this pull requests ping me and I'll add you as collaborator on my micropython fork.

My next step is to get my actual custom micropython firmware working using cmake.

I think I'll try to update or add directly to the micropython examples the options my firmware needs to build.

I need to get the c++ module to work and also a new example for a c or c++ module that interfaces to a static library.

stinos commented:
> Would the best place be the ESP32 port readme?

This is actually about user C modules so should probably go into docs/develop/cmodules.rst.

But also see #6960. There is quite some overlap, so is there are resuable bits between ports (apart from the usermod.cmake itself) that would best be made as such.

jej commented:
You should rename this pull request: "esp32: Create a mechanism to support user modules in cmake based builds." (look at the failed test)

mocleiri commented:
I rewrote the commits to tie into the rp2 port work for user modules.  It works for C modules but not for C++ modules.

This works
```
$ make USER_C_MODULES=/src/examples/usercmodule/cexample/

```

But building the cppexample directly fails:

```
$ make USER_C_MODULES=/src/examples/usercmodule/cppexample/

```

mocleiri commented:
I copied the line to generate the qstr.i.last file from the rp2 port over into the base py/mkrules.cmake file.  It works to generate a working firmware for the c and c++ example.  But I have no idea if this is correct or would impact the zeyphr cmake build that also depends on this file.

mocleiri committed:
ports/esp32: Restore USER_C_MODULE Support.

Support for User C and C++ modules was lost due to upgrading the esp32
to the latest CMake based IDF from the GNUMakefile build process.

Restore the support for the esp32 port by integrating with the approach
upstreamed by the rp2 port.

Signed-off-by: Michael O'Cleirigh <michael.ocleirigh@rivulet.ca>

mocleiri committed:
tools/ci.sh: Build User Modules for esp32.

Builds the esp32 port against the example C and CXX modules.

Signed-off-by: Michael O'Cleirigh <michael.ocleirigh@rivulet.ca>

mocleiri commented:
I have rebased my changes onto the latest from #6960 and it works to build both the C and C++ example.  I tested in both the espressif v4.0.2 and v4.3 idf's and it worked.    

I had some problem getting the **tools/ci.sh** script to run when trying to build esp32 with the **micropython/examples/usercomodule** example.

I tried the same relative path syntax which failed.

I tried assuming the current working directory was micropython which failed.

I set the absolute path to where the micropython/examples/usercmodule/micropython.cmake file is located and it worked.

mocleiri commented on tools/ci.sh:
```diff
@@ -100,6 +100,8 @@ function ci_esp32_build {
     make ${MAKEOPTS} -C ports/esp32 submodules
     make ${MAKEOPTS} -C ports/esp32
+    make ${MAKEOPTS} -C ports/esp32 clean
+    make ${MAKEOPTS} -C ports/esp32 USER_C_MODULES=/home/runner/work/micropython/micropython/examples/usercmodule/micropython.cmake
```
I tried to user ../../examples/usercmodule/micropython.cmake and examples/usercmodule/micropython.cmake here and both failed.  

I'm not sure why the behavior is different here for esp32 versus the other ports that are able to refer to the file in a relative manner.

dpgeorge replied:
It could be because of the location of the build set by `idf.py`.

The idea with this script/functions is that it can be executed locally, which won't work with this change.  Could you try `$PWD/../../examples/usercmodule/micropython.cmake` and also `$(pwd)/../../examples/usercmodule/micropython.cmake`?

dpgeorge replied:
See #7091 for a "fix" to this to use relative paths.  The issue was that the main CMakeLists.txt is in `ports/esp32/main/` so that's the directory from where relative paths are taken.

imliubo commented:
I try build ulab with this PR, but got an error,cexample module and cppexample no error.
```
make USER_C_MODULES=/home/imliubo/CodingBug/micropython/micropython-ulab/code/micropython.cmake
```
```bash
/home/imliubo/.espressif/tools/xtensa-esp32-elf/esp-2020r3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/main/libmain.a(mphalport.c.obj): in function `mp_hal_stdout_tx_str':
/home/imliubo/CodingBug/micropython/M5Stack-micropython/ports/esp32/mphalport.c:103: multiple definition of `mp_hal_stdout_tx_str'; esp-idf/main/libmain.a(stdout_helpers.c.obj):/home/imliubo/CodingBug/micropython/M5Stack-micropython/lib/utils/stdout_helpers.c:24: first defined here
/home/imliubo/.espressif/tools/xtensa-esp32-elf/esp-2020r3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/main/libmain.a(mphalport.c.obj): in function `mp_hal_stdout_tx_strn_cooked':
/home/imliubo/CodingBug/micropython/M5Stack-micropython/ports/esp32/mphalport.c:123: multiple definition of `mp_hal_stdout_tx_strn_cooked'; esp-idf/main/libmain.a(stdout_helpers.c.obj):/home/imliubo/CodingBug/micropython/M5Stack-micropython/lib/utils/stdout_helpers.c:14: first defined here
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/micropython.elf.dir/build.make:741: micropython.elf] Error 1
make[2]: *** [CMakeFiles/Makefile2:2085: CMakeFiles/micropython.elf.dir/all] Error 2
make[1]: *** [Makefile:130: all] Error 2
```

mocleiri committed:
ports/esp32: Fix multiple definition errors when building ulab.

It was noticed that the esp32 port didn't build ulab correctly.  The
problem was a multiple defintion of the 'mp_hal_stdout_tx_str' and
'mp_hal_stdout_tx_strn_cooked' functions.

They were defined in stdout_helpers.c but also in the
ports/esp32/mphalport.c.

I fixed the error by removing stdout_helpers.c from the build.

Signed-off-by: Michael O'Cleirigh <michael.ocleirigh@rivulet.ca>

mocleiri commented:
@dpgeorge I rebased this pull request onto master at (5976ea02a50ea76d72ada40db4cda186147fb412).  

I added a commit that fixes the linking issue when building ulab.  @imliubo can you retest on the latest?

```
(in git bash)
$ cd /c/git
$ git clone https://github.com/mocleiri/micropython.git
$ cd micropython
$ git checkout esp32-cmake-user-module-support
$ cd /c/git
$ git clone https://github.com/v923z/micropython-ulab.git
$  winpty docker run -i -t -v /$(pwd):/src espressif/idf:release-v4.3 bash
Adding ESP-IDF tools to PATH...
Using Python interpreter in /opt/esp/python_env/idf4.3_py3.6_env/bin/python
Checking if Python packages are up to date...
Python requirements from /opt/esp/idf/requirements.txt are satisfied.
Added the following directories to PATH:
  [ truncated ]
Done! You can now compile ESP-IDF projects.
Go to the project directory and run:

  idf.py build

root@3dae484d614d:/# cd /src/micropython/mpy-cross
root@3dae484d614d:/# make
root@3dae484d614d:/# cd /src/micropython/ports/esp32
root@3dae484d614d:/# make USER_C_MODULES=/src/micropython-ulab/code/micropython.cmake
...
exit docker container and flash /c/git/micropython/ports/esp32/build-GENERIC/firmware.bin to esp32 microcontroller.

```
I'm able to build and flash ulab using the espressif/idf:release-v4.3 idf docker image:
![image](https://user-images.githubusercontent.com/250942/113084974-0093fb80-91ad-11eb-8a00-23205e1a61e3.png)

imliubo commented:
Hi @mocleiri ,
The error is resolved, waiting for merge :)
```bash
>>> import ulab
>>> help(ulab)
object <module 'ulab'> is of type module
  __name__ -- ulab
  __version__ -- 2.5.1-2D
  dtype -- <function>
  numpy -- <module 'numpy'>
  scipy -- <module 'scipy'>
  utils -- <module 'utils'>
>>> 
```

dpgeorge commented:
Merged in 0ccd9e08aad95f955ded0006cff67e2a50778c84 through ec79e445027857d219d139cd31af578124a8e913 with minor modification to CI build to use a relative path.

Thank you @mocleiri for this, and for the others involved with testing etc!
